### ISSUE TITLE
[runtimes][3.8] AnimationState: attachment stays disabled after 2 transitions

### ISSUE THREAD

**Issue by HaraldCsaszar** [EsotericSoftware/spine-runtimes#1352]
Scenario (from #1352) is a->b->c where a and b key a slot's attachment to null, c does not key it, and the setup pose has an attachment shown.

Before the fix: a applies as if mixed in, which is needed in case a later animation keys the deform (fix from #1292). b is SUBSEQUENT so doesn't apply at all, leaving the attachment null after the mix. Expected: the attachment to be the setup pose after the mix.

After the fix: a applies as if mixed in, as before. b is SUBSEQUENT and is applied only if it is the last timeline, which it is in the scenario, so it sets the attachment to the setup pose.

[Repro files](http://esotericsoftware.com/files/issues/1352-attachment-timeline.zip). Skeleton has 3 animations, two of which hide a setup pose attachment. Third animation assumes setup pose attachment is set on slot. 

Commit: https://github.com/EsotericSoftware/spine-runtimes/commit/603c9e23668dde2b812262c6dbc0a22b1afa4d7a

- [x] spine-libgdx

---
- [x] spine-csharp
- [x] spine-xna
- [x] spine-monogame
- [x] spine-unity
- [x] spine-tk2d

---
- [x] spine-c
- [x] spine-cpp
- [x] spine-sfml
- [x] spine-cocos2d-objc
- [x] spine-cocos2dx
- [x] spine-ue4

---
- [x] spine-ts core
- [x] spine-ts webgl
- [x] spine-ts canvas
- [x] spine-ts threejs

---
- [x] spine-as3
- [x] spine-starling

---
- [x] spine-lua 
- [x] spine-love
- [x] spine-corona

## Old issue comment
As reported in the 3rd repro package in this email conversation:
FMfcgxwCgCRVHsmNCLDWbCVXZRXGLJGV

Reproduction asset:
https://esotericsoftware.com/files/issues/1352_Attachment-Stays-Disabled-Issue.zip
Contains exported skeleton assets called BBox.

Occurs on 3.8-beta branch with asset exported from Spine 3.8 only, could not be reproduced with 3.7 code and assets exported from Spine 3.7. Problem also occurs with 3.8-beta code and 3.7 assets.

**Context:**
The user has provided an exported skeleton where some animations have disabled/enabled different attachments during the whole animation (not switching in a keyframe in the middle of the animation).
Most animations have two `hand` attachments enabled (e.g. `Idle` and `Hit`), but the animation `DockedMusic2` has both `hands` disabled and has a `controller` attachment enabled instead of the hands.
The problem now occurs when switching from an animation with disabled `controller` attachment to another animation with disabled `controller` attachment and then, before MixDuration is over switching to an animation with enabled `controller` attachment (e.g. `DockedMusic2`).

**Problem reproduction steps:**
**In Skeleton Viewer**
1) Start the Skeleton Viewer with latest code from 3.8-beta branch.
2) Open the repro asset skeleton `BBox.json`.
3) Set `Default mix` to `0.2` (can also be reproduced when set to `0.3`)
4) Set `Speed` to `1.0`.
5) Select bottom animation in the list: `WalkDance` [disabled Controller]
6) Switch animations twice quickly by hitting the `down-arrow` key, this starts transitions to `DockedMusic` [disabled Controller] and then to `DockedMusic2` [enabled Controller].
7) Observe the still disabled controller attachment which should be displayed. Also observe the two correctly disabled hand attachments.
Note that the problem does not occur when switching from `DockedMusic2` [enabled Controller] to `DockedMusic` [disabled Controller] and back to `DockedMusic2` [enabled Controller] at same speed with arrow keys.
It also occurs when switching animations via the `up-arrow` key: starting at `Hit` [disabled Controller] to `Falling` [disabled Controller]  to `DockedMusic2` [enabled Controller].

**In Unity:**
1) Create a new empty project.
2) Import spine-unity runtime of 3.8-beta branch.
3) Import the `BBox` repro skeleton asset (first rename `BBox.atlas` to `BBox.atlas.txt`).
4) At the _SkeletonData asset, set (or better said leave) default mix duration to 0.2, no custom mix durations set.
5) Drag the skeleton to a new empty scene.
6) Create a new class `ReproAttachmentProblem.cs` in any Assets subdir with the following content:
```
public class ReproAttachmentProblem : MonoBehaviour
{
    [SerializeField]
    private SkeletonAnimation skeletonAnimation;

    IEnumerator Start()
    {
        skeletonAnimation.AnimationState.AddAnimation(0, "Idle", false, 0);
        skeletonAnimation.AnimationState.AddAnimation(0, "Hit", false, 1); // arbitrary time, at least 1 frame
        skeletonAnimation.AnimationState.AddAnimation(0, "DockedMusic2", true, 0.1f); // must be less than mix duration
        // now observe the incorrecly disabled controller attachment, both hands are also (correctly) disabled.
    }
}
```
7) Attach the reproduction component to the skeleton GameObject in the scene.
8) Hit play to watch the animations switching twice automatically, observe the incorrectly disabled controller attachment.

**Comment by badlogic** [comment]
This is a bug in `AnimationState`. The proposed fix let's the last attachment timeline for a specific slot always mix to the setup pose on mixing out. 

```
// if (!attachments && timeline instanceof AttachmentTimeline) continue;
if (!attachments && timeline instanceof AttachmentTimeline) {
    if ((timelineMode[i] & NOT_LAST) == NOT_LAST) continue;
    blend = MixBlend.setup;
}
```

[Repro files](http://esotericsoftware.com/files/issues/1352-attachment-timeline.zip). Skeleton has 3 animations, two of which hide a setup pose attachment. Third animation assumes setup pose attachment is set on slot.

**Comment by NathanSweet** [comment]
Write up for future Nate:

Scenario (from #1352) is `a->b->c` where `a` and `b` key a slot's attachment to null, `c` does not key it, and the setup pose has an attachment shown.

Before the fix: `a` applies as if mixed in, which is needed in case a later animation keys the deform (fix from #1292). `b` is `SUBSEQUENT` so doesn't apply at all, leaving the attachment null after the mix. Expected: the attachment to be the setup pose after the mix.

After the fix: `a` applies as if mixed in, as before. `b` is `SUBSEQUENT` and is applied only if it is the last timeline, which it is in the scenario, so it sets the attachment to the setup pose.
